flink-connector-mongodb-cdc/src/main/java/com/ververica/cdc/connectors/mongodb/MongoDBSource.java

```diff
@@ -0,0 +1,439 @@
+        }
+
+        /**
+         * errors.log.enable
+         *
+         * <p>Whether details of failed operations should be written to the log file. When set to
+         * true, both errors that are tolerated (determined by the errors.tolerance setting) and not
+         * tolerated are written. When set to false, errors that are tolerated are omitted.
+         */
+        public Builder<T> errorsLogEnable(boolean errorsLogEnable) {
+            this.errorsLogEnable = errorsLogEnable;
+            return this;
+        }
+
+        /**
+         * errors.tolerance
+         *
+         * <p>Whether to continue processing messages if an error is encountered. When set to none,
+         * the connector reports an error and blocks further processing of the rest of the records
+         * when it encounters an error. When set to all, the connector silently ignores any bad
+         * messages.
+         *
+         * <p>Default: "none" Accepted Values: "none" or "all"
+         */
+        public Builder<T> errorsTolerance(String errorsTolerance) {
+            this.errorsTolerance = errorsTolerance;
+            return this;
+        }
+
+        /**
+         * heartbeat.interval.ms
+         *
+         * <p>The length of time in milliseconds between sending heartbeat messages. Heartbeat
+         * messages contain the post batch resume token and are sent when no source records have
+         * been published in the specified interval. This improves the resumability of the connector
+         * for low volume namespaces. Use 0 to disable.
+         */
+        public Builder<T> heartbeatIntervalMillis(int heartbeatIntervalMillis) {
+            checkArgument(heartbeatIntervalMillis >= 0);
+            this.heartbeatIntervalMillis = heartbeatIntervalMillis;
+            return this;
+        }
+
+        /**
+         * The deserializer used to convert from consumed {@link
+         * org.apache.kafka.connect.source.SourceRecord}.
+         */
+        public Builder<T> deserializer(DebeziumDeserializationSchema<T> deserializer) {
+            this.deserializer = deserializer;
+            return this;
+        }
+
+        /** Build connection uri. */
+        private ConnectionString buildConnectionUri() {
+            StringBuilder sb = new StringBuilder(MONGODB_SCHEME).append("://");
+
+            if (username != null && password != null) {
+                sb.append(encodeValue(username)).append(":").append(encodeValue(password));
+            }
+
+            sb.append("@").append(checkNotNull(hosts));
+
+            if (StringUtils.isNotEmpty(connectionOptions)) {
+                sb.append("/?").append(connectionOptions);
+            }
+
+            return new ConnectionString(sb.toString());
+        }
+
+        /**
+         * The properties of mongodb kafka connector.
+         * https://docs.mongodb.com/kafka-connector/current/kafka-source
+         */
+        public DebeziumSourceFunction<T> build() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.